geoh5py/shared/entity_type.py

```diff
@@ -20,136 +20,209 @@
 import uuid
 import weakref
 from abc import ABC, abstractmethod
-from typing import TYPE_CHECKING, TypeVar, cast
+from typing import TYPE_CHECKING, Any, TypeVar, cast
 
 if TYPE_CHECKING:
-    from .. import workspace as ws
-
+    from ..workspace import Workspace
 
 EntityTypeT = TypeVar("EntityTypeT", bound="EntityType")
 
 
 class EntityType(ABC):
-    _attribute_map = {"Description": "description", "ID": "uid", "Name": "name"}
+    """
+    The base class for all entity types.
 
-    def __init__(self, workspace: ws.Workspace, uid: uuid.UUID | None = None, **kwargs):
-        assert workspace is not None
-        self._workspace = weakref.ref(workspace)
+    :param workspace: The workspace to which the entity type belongs.
+    :param uid: The unique identifier of the entity type.
+    :param description: The description of the entity type.
+    :param name: The name of the entity type.
+    :param kwargs: Additional keyword arguments to set as attributes.
+    """
 
-        assert uid is None or isinstance(uid, uuid.UUID)
-        self._description: str | None = "Entity"
-        self._name: str | None = "Entity"
-        self._on_file = False
-        self._uid: uuid.UUID = uid if uid is not None else uuid.uuid4()
+    # todo: This mechanic feels quite hacky
+    _attribute_map = {"Description": "description", "ID": "uid", "Name": "name"}
 
-        for attr, item in kwargs.items():
-            try:
-                if attr in self._attribute_map:
-                    attr = self._attribute_map[attr]
-                setattr(self, attr, item)
-            except AttributeError:
-                continue
+    def __init__(
+        self,
+        workspace: Workspace,
+        uid: uuid.UUID | None = None,
+        description: str | None = "Entity",
+        name: str | None = "Entity",
+        **kwargs,
+    ):
+
+        self._workspace: Workspace = self._set_workspace(workspace)
+        self.uid = self._modify_attribute("ID", uid, **kwargs)
+        self.description = self._modify_attribute("Description", description, **kwargs)
+        self.name = self._modify_attribute("Name", name, **kwargs)
```

Comment:
It would probably make more sense to do the translation directly on the reader, and only deal with formatted arguments here, and send it straight to the setters

Reply:
agree

Reply:
I will put that in find or create